### DISCUSSION TITLE
How to change config-directory-path for internal commands?

### DISCUSSION THREAD

**Discussion by theRoboxx** [leo-arch/clifm#297] [Q&A]
Hi there,
I'm using NixOS and build my system via my config files.
Therefore the config files for clifm are stored under `~/nixos-config/config/clifm/` and not `~/.config/clifm/`.
Is it possible to change this path for internal command like `config` or `bm add`, so it uses the files inside `~/nixos-config/config/clifm/profiles/default/clifmrc` or `~/nixos-config/clifm/profiles/default/cbookmarks.clifm`?

Thanks in advance!

**Comment by leo-arch** [comment]
Hi  @theRoboxx. Try with `-D,--config-dir`.

```sh
clifm -D ~/nixos-config/clifm
```

Take a look also at `-b,--bookmarks-file` and `-c,--config-file`.

**Reply by theRoboxx** [reply]
- `~/.config/clifm/` is (1)
- `~/nixos-config/config/clifm/` is (2) 

Thanks, I saw the options but would like to leave the directory (1) as my config directory that gets loaded for clifm.
Instead, I would prefer to make the changes permanent only after I deploy them within NixOS.
So **not immediately**, as if I were to do with `clifm -D (2)`. 
This has the advantage that I can easily rollback the deployed changes if something breaks.

tl;dr for proposed config setting:
- (1) stays the config directory
- change only the directory-path for the commands like `config` -> `(2)/clifmrc`, `bm add` -> `(2)/bookmarks.clifm`
    - something like `--config-dir-internal-commands=DIR`

I know, this sound like a niche use case. 
But it could be useful for all immutable OS's in the future.

**Reply by leo-arch** [reply]
Wouldn't `clifm -b ~/nixos-config/config/clifm/bookmarks.clifm -c ~/nixos-config/config/clifm/clifmrc` do the trick?

**Reply by theRoboxx** [reply]
Unfortunately no, because clifm should reference the config from the `nix/store` that gets linked after `sudo nixos-rebuild switch`.
Lets say I accidentally delete/alter my clifmrc. 
With your current approach, I can't use an old build because I load it to `.config/clifm` and there it gets linked from the `nix/store` which holds all versions for every rebuild. So this way I would be stuck with the broken/missing config file(s).

Left is the config from the rebuild and right is the manual edited version.
I can rollback the left version via NixOS but not the right version that easily.
![hypr2](https://github.com/leo-arch/clifm/assets/63428105/a2e3a09f-b4c8-49f1-898b-a457eb1f0f83)

The files get linked from the `nix/store`, so If I select the build from yesterday it chooses the automatically-stored files from yesterday and not the recent edited files from now.
![hypr1](https://github.com/leo-arch/clifm/assets/63428105/6f5dcd7a-e219-4460-8040-6889cc8cc2ea)

I hope this explanation makes sense.

**Reply by leo-arch** [reply]
> I hope this explanation makes sense.

It surely does, but I still don't quite get it. Is there some program already implementing this feature, so that I can take it as a model?

For example, you say:

> So not immediately, as if I were to do with clifm -D (2)

If not immediately (i.e. at runtime), when? You add: `only after I deploy them within NixOS`. But I don't know what this means.

**Reply by leo-arch** [reply]
Let me see if I understand the procedure, which, based on what you say (`If I select the build from yesterday it chooses the automatically-stored files from yesterday and not the recent edited files from now`), I reconstruct as follows:

1. You create a "snapshot" (don't know if this is the appropriate word) of the system (including current **clifm**'s config files) via `nixos-rebuild switch` (or some other command, or maybe automatically, it doesn't matter). Config files are stored in `/nix/store` (or somewhere else, it doesn't matter either).
2. You modify some **clifm** settings, altering thus the config files. These changes are not reflected in the snapshot taken in step 1 (because modifications are made rather on `~/.config/clifm`).
3. You want to rollback to the state the system was during the snapshot taken in step 1. Thereby, you lose the changes made in step 2 (but you want to keep them).

Is this correct? If so, I still don't see what **clifm** can do to help. I cannot but feel that a little script executed before (or after) **clifm**, automatically updating config files in `/nix/store` (or wherever), would be the simplest and most intuitive solution. But, doesn't this defeat the idea of an immutable OS?

**Reply by theRoboxx** [reply]
I think, I over-complicated it a bit, sorry.
This is what I would like to be able to do:

---
10. `.config/clifm/profiles/default/clifmrc` is used as the config file when starting clifm.
20. The internal command `config` should open the the `bla/xxx/yyy/file` file for editing.

30.`.config/clifm/profiles/default/bookmarks.clifm` is used as the bookmarks file when starting clifm.
40. The internal command `bm add` should add bookmarks to the `bla/xxx/yyy/book.clifm` file.

---
Currently the `-c` flag changes the path for (10) **and** (20).
The `-b` flag changes the path for (30) **and** (40).

**Reply by leo-arch** [reply]
But, if (10) is loaded at startup, and `config` modifies (20), the next time you start **clifm** changes made to (20) before will be lost, because you will be loading again from (10). Same thing with bookmarks. What's the point? Why do you want to add a bookmark that won't survive the current session?

**Reply by theRoboxx** [reply]
Ok, now I see what is missing.

In NixOS you have a dedicated directory for all the configs in the OS and applications.
In my example I choose `~/nixos-config/` as the directory.
**If I build the system, everything from `~/nixos-config/` gets stored first in `/nix/store` as a new Snapshot. From there I need to map all config files and where they should be linked for the applications to use them, like clifm.**

```
    ".config/clifm/profiles/default/clifmrc" = {
      source = ~/nixos-config/config/clifm/profiles/default/clifmrc;
    };
    ".config/clifm/profiles/default/bookmarks.clifm" = {
      source = ~/nixos-config/clifm/profiles/default/bookmarks.clifm;
    };

```
I can't link them directly from `/nix/store`  to `~/nixos-config/clifm/profiles/default/.` because the version I edit and the version from `nix/store` are different and therefore I need (10) and (20) to have a different path. 
(10) should be `.config/clifm/profiles/default/clifmrc` and (20) should be `~/nixos-config/config/clifm/profiles/default/clifmrc`.

`nix/store` is managed by the os and only gets changed by creating/deleting snapshots.

**Reply by leo-arch** [reply]
Thanks for the info @theRoboxx . I'll keep it in mind and let you know if anything comes out.
